README.md

```diff
@@ -1,3 +1,88 @@
 # ioc-hierarchy-service
 In-memory queryable hierarchy service
 
+This project generates a debian package with all components necessary to
+create an in-memory cache of a hierarchy. The hierarchy is modelled as
+a graph and can be queried in a number of ways. The cache itself is run
+as a daemon process which can be backed up by using another daemon
+process called backup service. All interprocess communication is handled
+with ZeroMQ.
+
+Querying the process must be done using the bundled python module. This
+can be imported and a function named `query` can be run, e.g. to check
+the health of the service:
+
+```
+>>> import tag_hierarchy_client
+>>> tag_hierarchy_client.query([{'command': 'healthcheck'}], "server")
+```
+
+# Architectural overview
+The following diagram describes which components are bundled with the
+debian package and how they are intended to be used:
+
+![Architecture diagram][docs/hierarchy_daemon.png]
+
+## Communication
+Components communicate via ZeroMQ(ZMQ). We use the patterns REQ-REP
+(request-response) when the python client communicates with the server
+or the statemanager.
+
+From the python perspective, a request has to pass a list of dictionaries
+and expects a list of dictionaries in return. By convention the first
+dictionary of a request constitutes the command. Any number of
+subsequent dictionaries constitute the data associated with the command.
+
+The python module is built in C++ using the package `pybind11`, which
+translates the python dictionaries into C++ native `std::map`. Since
+C++ is statically typed, only a subset of dictionary key and value
+types are permitted, and this is supported on the C++ by using
+`boost::variant`. Currently the supported dictionary type is
+`Dict[str, Union[bool, int, float, none, str, List[int], List[str]]]]`
+Trying to pass objects that do not conform will raise a `TypeError`
+
+## Tag hierarchy service
+The tag hierarchy service keeps a tag hierarchy in memory. To populate
+the tag hierarchy, use the `populate_hierarchy` command. To query the
+hierarchy, the `nodes` command is used.
+
+### Flow
+Populating the hierarchy can take quite a bit of time, therefore the
+package supports running in a statemanager mode which should not be used
+for other purposes than managing the state of the hierarchy.
+
+When the cache is updated, the population process happens against the
+statemanager while the server is still serving requests. When the
+population is done, it is asked to store itself via the backup service.
+When the backup service has stored a new state, it notifies the server
```

Comment:
The "server" here is the user of the python client?

Reply:
It is the tag-hierarchy server, it listens to new updated states from the backup service. When a new update has been posted, the server is notified so that it can fetch it and initialize itself with a new state. Can try to state it more clearly

Reply:
ah. ok. there seems to be three "servers" here. tag-hierarchy server, statemanager server and backup server.